staubli_val3_driver/val3/ros_server/interpolateVel.pgx

```diff
@@ -0,0 +1,21 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Programs xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.staubli.com/robotics/VAL3/Program/2">
+  <Program name="interpolateVel">
+    <Parameters xmlns="http://www.staubli.com/robotics/VAL3/Param/1">
+      <Parameter name="x_mMotion" type="MotionCmd" xsi:type="element" />
+    </Parameters>
+    <Code><![CDATA[begin
+  // Applies velocity "smoothing" to last point in trajectory when vel==0 is replaced by vel==0.1 (default for MoveIt)
+
+  // Round to remove trailing decimal places from fromBinary()
+  x_mMotion.mDesc.vel = x_mMotion.mDesc.vel * 10000
+  x_mMotion.mDesc.vel = round(x_mMotion.mDesc.vel)
+  x_mMotion.mDesc.vel = x_mMotion.mDesc.vel / 10000
+  // Smooth last point in traj
+  if((x_mMotion.mDesc.vel == 0.1) and (x_mMotion.nSequence != 0))
+    x_mMotion.mDesc.vel = (nVel[0] - (nVel[1] - nVel[0]))
```

Comment:
@gavanderhoorn @marshallpowell97: Couldn't this end up in a negative velocity if `nVel[0] == 0.0`? And if so, what would the controller do? Would it **skip** the point completely?

I am asking because I see a concrete problem on our setup: after I cherry picked this change to my private fork (*) I started noticing that the very last point of some trajectories is never reached, at all. I am not sure this change causes it, however, it is the main suspect at the moment.

(*disclaimer: my fork is not identical, however, the joint trajectory execution part should be untouched and matching what you have here)

Reply:
Technically it could become negative.

In practive however, iirc, 0 velocities are not used by the nodes in the IRC. Are you using a custom client as well?

Reply:
@marshallpowell97: could you comment on what the controller would do in that case?

Reply:
>  Are you using a custom client as well?

I am using the industrial robot client (IRC) with the modifications for cartesian trajectories I showed you long ago. In principle, I haven't changed anything on the joint trajectory handling thou, and we've been used this modified IRC client for many months and this behavior only started occurring recently. After going through changes applied to the setup, this stands out as the most likely candidate, however, I cannot 100% rule out something else.

Reply:
Well, I'm not claiming the code is faultless. Could you perhaps share a `.bag`/wireshark capture of a trajectory which exhibits the problem you're encountering?

Reply:
Yep, I will capture some data and post it here later today. I'm also waiting for an SRS license dongle to be shipped home (now that we're not allowed in the building) to test this on simulation, but it would be great to hear from @marshallpowell97 how the controller is supposed to behave on negative velocities.

Reply:
>  Couldn't this end up in a negative velocity if nVel[0] == 0.0? And if so, what would the controller do? Would it skip the point completely?

That would cause a runtime error. But a velocity of 0 would also cause a runtime error.

> I started noticing that the very last point of some trajectories is never reached, at all.

Hmmm that's pretty strange. Changing the velocity shouldn't cause the controller to not reach the last point, it would just slow down.

Testing on my machine, I don't see a difference in accuracy with and without the changes. Both trajectories seem to reach their final point (+/- 0.01deg) although I am only moving to a joint position (all joints = 0deg). @gavanderhoorn  Are there any tutorials I can follow to create a trajectory to a cartesian position so I can test? @gonzalocasas If you have a way for me to replicate your trajectory I can simulate it and/or run on a real robot.

Reply:
@marshallpowell97 the trajectory were I consistently see this problem is the here (in JSON format): https://gist.github.com/gonzalocasas/a52a63b579d8a1d8fc4ad983a319b39d

I will prepare a rosbag soon as well, in case that is easier. The trajectory was calculated with MoveIt in a pretty standard setup. The joints are prefixed because we have 2 arms (TX2-60l) in the model.

Reply:
This is not really a format @marshallpowell97 can directly use.

@gonzalocasas: it would be very much appreciated if you could convert this into something like [this](https://github.com/gavanderhoorn/motoman/blob/50b9fd428758932beb32fa60fa9e27a4ffba1214/motoman_driver/examples/simple_trajectory_action_client.py). That would make reproduction much easier.

Reply:
I would also suggest to post both a `.bag` (with the trajectory topics as well as `/joint_states`) and a wireshark capture, so we have both the ROS side and the Simple Message side and can compare.

Reply:
Finally: let's open a new issue, as a comment on a already merged PR is not very well suited for this.

Reply:
I added the rosbag file to the gist (https://gist.github.com/gonzalocasas/a52a63b579d8a1d8fc4ad983a319b39d#file-unfinished-trajectory_2020-03-20-14-57-09-bag). The robot is namespaced under `/microphone/`. The trajectory is sent directly to the service interface (`joint_path_command`) so I guess it does not show up in this bag file.

Reply:
> Finally: let's open a new issue, as a comment on a already merged PR is not very well suited for this.

Agree! Just created a new issue here https://github.com/ros-industrial/staubli_val3_driver/issues/16  (I started off as comment, because I am unsure about the source of this, but now it is long enough to be an issue on its own)